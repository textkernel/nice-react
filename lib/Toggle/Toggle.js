import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import niceProps from '../props';

class Toggle extends Component {
    constructor(props) {
        super(props);

        const { on } = this.props;

        this.state = {
            on
        }
    }

    componentWillReceiveProps(nextProps) {
        const { on } = nextProps;

        if (on === null) {
            return false;
        }

        this.setState({
            on
        });

        return true;
    }

    isControlled() {
        const { on } = this.props;
        return typeof(on) === typeof(true);
    }

    render() {
        const { children, className, context, disabled, id, offMuted, defaultOn, onChange, small, style } = this.props;
        const { on } = this.state;

        return (<span>
            <input
                type="checkbox"
                className={ classNames('nice-toggle', {
                    [context]: context
                }) }
                id={ id }
                disabled={ disabled }
                defaultChecked={ !this.isControlled() ? defaultOn : null }
                checked={ this.isControlled() ? on : null }
                onChange={ (e) => {
                    if (!onChange || disabled) {
                        return false;
                    }

                    const { checked } = e.target;
                    this.setState({
                        on: checked
                    });

                    return onChange(checked);
                }}
            />
            <label
                className={ classNames('nice-toggle-btn', {
                    small,
                    'off-muted': offMuted
                }, className) }
                style={ style }
                htmlFor={ id }
            >
                { children }
            </label>
        </span>);
    }
}

Toggle.propTypes = {
    id: PropTypes.string.isRequired,
    children: PropTypes.oneOfType([
        PropTypes.array,
        PropTypes.node
    ]),
    className: PropTypes.string,
    context: PropTypes.oneOf(niceProps.toggleContexts),
    disabled: PropTypes.bool,
    offMuted: PropTypes.bool,
    defaultOn: PropTypes.bool,
    on: PropTypes.bool,
    onChange: PropTypes.func,
    small: PropTypes.bool,
    style: PropTypes.object
};

Toggle.defaultProps = {
    children: null,
    className: null,
    context: null,
    disabled: false,
    offMuted: false,
    defaultOn: false,
    on: null,
    onChange: null,
    small: false,
    style: {}
};

export default Toggle;
