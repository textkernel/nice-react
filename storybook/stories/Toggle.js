import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, text, select } from '@storybook/addon-knobs';
import { withInfo } from '@storybook/addon-info';
import props from '../../lib/props';

import { Toggle } from 'nice-react';

const stories = storiesOf('Toggle', module);

stories.addDecorator((story, context) => withInfo('')(story)(context));
stories.addDecorator(withKnobs);

stories.add('Custom properties', () => {
    const contextOptions = props.toggleContexts;
    const context = select('Context', contextOptions, null);
    const disabled = boolean('Disabled', false);
    const on = boolean('On', false);
    const small = boolean('Small', false);
    const label = text('Label', '');
    const offMuted = boolean('Muted label when switched off', false);

    return <Toggle context={ context } disabled={ disabled } on={ on } offMuted={ offMuted } small={ small }>
        { label }
    </Toggle>;
});
