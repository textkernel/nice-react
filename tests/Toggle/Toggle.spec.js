import React from 'react';
import Toggle from '../../lib/Toggle';

describe('<Toggle />', () => {
    it('should render a simple Toggle', () => {
        const wrapper = shallow(<Toggle>Label</Toggle>);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render a disabled Toggle', () => {
        const wrapper = shallow(<Toggle disabled>Disabled toggle</Toggle>);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render a small primary Toggle that is on by default', () => {
        const wrapper = shallow(<Toggle context="primary" small on>Label</Toggle>);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render a toggle with custom class / style + muted label when switched off', () => {
        const wrapper = shallow(<Toggle className="custom-class" style={{ visibility: 'hidden' }} offMuted>Muted label</Toggle>);
        expect(wrapper).toMatchSnapshot();
  });
});
