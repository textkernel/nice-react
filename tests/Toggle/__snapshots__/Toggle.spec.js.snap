// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> should render a controlled Toggle that is off by state 1`] = `
<span>
  <input
    checked={false}
    className="nice-toggle primary"
    defaultChecked={null}
    disabled={false}
    id="myToggle"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="nice-toggle-btn small"
    htmlFor="myToggle"
    style={Object {}}
  >
    Label
  </label>
</span>
`;

exports[`<Toggle /> should render a disabled Toggle 1`] = `
<span>
  <input
    checked={null}
    className="nice-toggle"
    defaultChecked={false}
    disabled={true}
    id="myToggle"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="nice-toggle-btn"
    htmlFor="myToggle"
    style={Object {}}
  >
    Disabled toggle
  </label>
</span>
`;

exports[`<Toggle /> should render a simple Toggle 1`] = `
<span>
  <input
    checked={null}
    className="nice-toggle"
    defaultChecked={false}
    disabled={false}
    id="myToggle"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="nice-toggle-btn"
    htmlFor="myToggle"
    style={Object {}}
  >
    Label
  </label>
</span>
`;

exports[`<Toggle /> should render a small primary Toggle that is on by default 1`] = `
<span>
  <input
    checked={null}
    className="nice-toggle primary"
    defaultChecked={true}
    disabled={false}
    id="myToggle"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="nice-toggle-btn small"
    htmlFor="myToggle"
    style={Object {}}
  >
    Label
  </label>
</span>
`;

exports[`<Toggle /> should render a toggle with custom class / style + muted label when switched off 1`] = `
<span>
  <input
    checked={null}
    className="nice-toggle"
    defaultChecked={false}
    disabled={false}
    id="myToggle"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="nice-toggle-btn off-muted custom-class"
    htmlFor="myToggle"
    style={
      Object {
        "visibility": "hidden",
      }
    }
  >
    Muted label
  </label>
</span>
`;
